### ISSUE TITLE
perf: 🚀 Analyse browser rendering time in Firefox

### ISSUE THREAD

**Issue by SchulteMarkus** [solid-design-system/solid#1201]
# Current behavior

At https://www.union-investment.de/, I have noticed that the browser performance varies between Firefox and Chrome.
The rendering in Firefox is noticeably slower.  The difference in rendering time becomes even more visible when caching is deactivated in the developer tools (I suspect that there are relatively many new visitors to union-investment.de).

https://github.com/solid-design-system/solid/assets/1948684/0264a340-5a55-4c62-a8cd-944475b0bafb

In this example video you see a visit to https://www.union-investment.de/ with disabled caching. Firefox load time is 4.6s, while Chrome has 0.5s.

# Steps to reproduce

Visit https://www.union-investment.de/ using Chrome and Firefox.

# Benefits

[A good performance is important for visitors](https://web.dev/learn/performance/why-speed-matters)

# Technical Information

- OS: Windows 11
- Chrome v126
- Firefox developer edition v129, Browser Privacy: Standard

## DoR
- [ ] Item has business value
- [ ] Item has been estimated by the team
- [ ] Item is clear and well-defined
- [ ] Item dependencies have been identified

## DoD
- [ ] It was investigated whether Solid has a share in this situation
- [ ] If this is not the case, the presumed originators were made aware of this fact
- [ ] If this is the case, it was analysed whether possible measures are proportionate
- [ ] If this is the case, follow-up issue(s) haven been created

**Comment by mariohamann** [comment]
Thanks for raising this issue. There are two main issues here, before we could investigate, what Solid can do here:
1. The whole frontend of `union-investment.de` is generated with Web Components. The CMS (and any other application at Union Investment) would have to utilize SSR a lot more, instead of doing everything on the Client's side.
2. The Web Components utilize Stencil's Lazy Loading, which on the hand means, that only the components are loaded, that are needed (reduced amount of KB), but on the other hand leads to the fact, that ES Modules are loaded in parallel and sometimes delayed. This is very extreme for the navigation items, which load 200ms later then the rest of the components:
![CleanShot 2024-07-12 at 08 15 46](https://github.com/user-attachments/assets/a411492f-d81d-458a-ba6b-e7589cb36760)

The whole thing comes evident when loading a Solid Component from Server Side:
- 150ms for first Solid Component – 
![CleanShot 2024-07-12 at 08 26 16](https://github.com/user-attachments/assets/5130bad4-940b-4ce5-8aa2-64d98260312b)
- 1000ms until all CMS Components are loaded – ![CleanShot 2024-07-12 at 08 25 42](https://github.com/user-attachments/assets/c71c44e7-3ffb-40a6-bf66-72e05a46fa05)

**Comment by mariohamann** [comment]
I hope it's fine to close the ticket. If you have further questions or insights, feel free to re-open or come back anytime.

**Comment by SchulteMarkus** [comment]
The ultimate performance of Solid is certainly a challenging topic, and the performance on https://www.union-investment.de/ is certainly also dependent on the CMS. 
Your feedback, @mariohamann, deals more with general statements about performance. 
This issue is about the difference in performance between Firefox and Chrome (the CMS makes no difference here). I think it's worth looking into why the performance in Firefox is worse than in Chrome - even if all that comes out is that the Firefox engine is slower.

**Comment by mariohamann** [comment]
It's less dependent on the CMS but how the CMS project works (lots of little micro frontends which are hydrated on runtime).

From what you shared I can't see if Solid is responsible for the problems. Actually my test form before on FireFox shows:

![CleanShot 2024-07-12 at 16 30 40](https://github.com/user-attachments/assets/4e28be39-6c9d-444e-8a6c-a2e972a8cd18)

After the start (1), Solid is already running after some milliseconds (2), and 5 seconds later the CMS components are done. But as said – the latter has nothing to do with Solid. We have to investigate further if you see concrete problems, but from what I see it's more an architectural thing of the CMS project instead of Solid.

**Comment by SchulteMarkus** [comment]
I brought this up because I noticed this big difference at InvestmentWelt investmentwelt.de.
The CMS is not used at all in the imprint there. Here too there is a difference in the performance of Firefox compared to Chrome, specifically I measure a difference of around 1 second for "Finish". Of course, solid is not the only component that affects performance here either.
Is there an environment in which the performance of Solid of Firefox compared to Chrome can be analysed in isolation?

**Comment by mariohamann** [comment]
I'm not sure, where your theory is coming from, that there might be a problem with Solid and FireFox, but okay, here we are:

Methodology: I created a local environment with a TON of `<sd-*>` elements 
`<sd-*>` elements are a wild mixture form our docs, being buttons, inputs, teasers, dropdowns, teasers, carousels, accordions etc.. for comparison: `union-investment.de` seems to use about 96 elements. I measured roundabout the time from nothing until this screen is loaded:

![CleanShot 2024-07-12 at 21 17 01@2x](https://github.com/user-attachments/assets/5d4451ed-b169-4823-b77d-46ef803234fb)

| Elements                     | FireFox | Edge   |
|------------------------------|---------|--------|
| 635 `<div>`s, 844 `<sd-*>`   | 363ms   | 710ms  |
| 1905 `<div>`s, 2532 `<sd-*>` | 999ms   | 2520ms |

Again – it's not about the usage of the CMS in the background (which just serves the HTML) but about the Frontend that is completely generated by JavaScript – and I expect that this is the same in investmentwelt.de

I would stop at this point here, but feel free to come up with own measurements, that could lead us into more concrete performance investigations – we're keen on optimizing our library and therefore are very open for insights!

**Comment by SchulteMarkus** [comment]
Thanks for working through the differences in browser performance @mariohamann. Exciting result, I think, and you explain well why Solid is probably not responsible for the observed differences.
